### ISSUE TITLE
'IoTHubDeviceClient_CreateFromDeviceAuth API' will return non-NULL value in failure scenario(Internet is disconnected)) 

### ISSUE THREAD

**Issue by ykgw-moorthy** [Azure/azure-iot-sdk-c#1968]
**Development Machine, OS, Compiler (and Other Relevant Toolchain Info)**

Compiled on  Windows10 Enterprise with MSVC. compiler

**SDK Version (Please Give Commit SHA if Manually Compiling)**

LTS_01_2021_Ref01
Release on Jan 23

**Protocol**

MQTT

**Describe the Bug**
When **IoTHubDeviceClient_CreateFromDeviceAuth** API is called **without internet connection**, it is expected  return invalid device handle with NULL value. But currently it is returning device handle with non-NULL value.

Based on the below documentation it is expected to return NULL value when failed to establish the cloud connection(create valid device handle).

**https://docs.microsoft.com/en-us/azure/iot-hub/iot-c-sdk-ref/iothub-device-client-h/iothubdeviceclient-createfromdeviceauth**

```
if (reinitialization_required==true)
{
protocol_prov=MQTT_Protocol;
device_handle=NULL;
if ((device_handle = IoTHubDeviceClient_CreateFromDeviceAuth(iothub_uri, device_id, protocol_prov)) == NULL)
{
printf("Failed to create the device handle");
}
else
{
IoTHubDeviceClient_Destroy(device_handle);
}

}

```
Please note that in above example we are setting values for iothub_uri and device_id values during the registering the device(Prov_Device_Register_Device).

In the above example, when internet connection disabled, IoTHubDeviceClient_CreateFromDeviceAuth API will return non NULL value. If we try to destroy this device handle after creation will lead to exception and crash the executable.

Basically we are trying to validate the device handle once it is created by comparing it with NULL value. But we observed that it is returning non NULL value. Please let us know if there is better approach to validate the device handle once it is created.

**Comment by danewalton** [comment]
Hi @ykgw-moorthy can I ask what exactly you're trying to validate here? Are you expecting that if the handle is not `NULL` then the connection is established? If so that's not the intention of the function's return value and I don't see in the doc you linked where it says that. Just to be clear

> Based on the below documentation it is expected to return NULL value when failed to establish the cloud connection(create valid device handle).

A valid device handle and a successful connection are not the same thing here.

For connection information I would recommend using the [connection status callback](https://docs.microsoft.com/en-us/azure/iot-hub/iot-c-sdk-ref/iothub-device-client-h/iothubdeviceclient-setconnectionstatuscallback) to understand the state of the connection.

Lastly, If you are indeed getting a crash when calling `_Destroy` that could be worrisome. Could you share more info on where that might be crashing?

**Comment by ykgw-moorthy** [comment]
Thanks @danewalton  for your response.
It was my bad!! I had assumed that if there is no internet while calling  IoTHubDeviceClient_CreateFromDeviceAuth API, it will not create valid device handle.  I had assumed that it will establish connection with AZURE during device handle creation. so I was expecting the null device handle. Thanks for confirmation

Regarding crash, following is screenshot of exception. I am getting crash in _iothubtransport.c_ file.

![image](https://user-images.githubusercontent.com/80386183/120677584-da445200-c4b4-11eb-8349-8222b478e3f3.png)

This exception is observed when we call `_Destroy` API immediately after creation of device handle(IoTHubDeviceClient_CreateFromDeviceAuth ).

If we call `_Destroy` API with a delay of around 30 seconds, I do not see this exception.

**Comment by ericwol-msft** [comment]
The only why this can happen if you call a IoTHubClient_XXX API with a iotHubClientHandle after closing the handle. A callstack would be useful.

**Comment by ykgw-moorthy** [comment]
Yes. It was mistake in my implementation. Thanks for support
